File: 0629-K Inverse Pairs Array/0629-K Inverse Pairs Array.cpp
class Solution {
public:
    int kInversePairs(int n, int k) {
        vector<int> presum(1 + k);
        int MOD = 1000000007;
        for (int i = 1; i <= n; ++i)
        {
            vector<int> temp(1 + k);
            temp[0] = 1;
            for (int j = 1; j <= k; ++j)
            {
                int val = (presum[j] - ((j - i >= 0) ? presum[j - i] : 0) + MOD) % MOD;                
                temp[j] = (temp[j - 1] + val) % MOD;
            }
            presum = temp;
        }
        
        return (presum[k] - ((k > 0) ? presum[k - 1] : 0) + MOD) % MOD;
    }
};

File: 0601-0700/0629-K Inverse Pairs Array/0629-K Inverse Pairs Array.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        presum = [0] * (1 + k)
        MOD = 1000000007
        for i in range(1, 1 + n):
            temp = [0] * (1 + k)
            temp[0] = 1
            for j in range(1, 1 + k):
                val = (presum[j] - (presum[j - i] if j >= i else 0)) % MOD
                temp[j] = (temp[j - 1] + val) % MOD
            presum = temp
        return (presum[k] - (presum[k - 1] if k > 0 else 0)) % MOD
